lang/wasm_alien_thread_runner.cc

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/reactor.hh>
+
+#include "lang/wasm.hh"
+#include "lang/wasm_alien_thread_runner.hh"
+
+using namespace std::chrono_literals;
+
+namespace wasm {
+
+std::optional<wasm_compile_task> task_queue::pop_front() {
+    if (_sem.try_acquire_for(10ms)) {
```

Comment:
This just generates unnecessary wakeups. Better to use a condition variable and a mutex, then you can wake up the thread for when something is queued or when _stopped is set. Another option is to have a queue of std::optionals, and stop when nullopt is queued.

Reply:
yeah, the reason why `ShardedWorkQueue` uses `try_acquire_for()` is get it waked up for checking if the thread pool is stopped or not. in our case, an `std::optional<>` or `std::unique_ptr<>` would serve the purpose better. as Avi puts, the nullopt or an empty pointer would be a `sentinel`. i'd suggest use an `optional<>`, as `task_queue::pop_front()` already returns an `optional<>`.

Reply:
Ok, I tried to use the approach with a condition variable and an `optional` sentinel in the rebase